File: 198-house-robber/house-robber.java
class Solution {
    int[] dp;
    int dfs(int cur,int[] nums){
        if(cur >= nums.length)
            return 0;
        
        if(dp[cur] != -1)
            return dp[cur];
        
        int rob = nums[cur] + dfs(cur+2,nums);
        int noRob = dfs(cur+1, nums);
        dp[cur] = Math.max(rob,noRob);

        return dp[cur];
    }
    public int rob(int[] nums) {
        dp = new int[nums.length];
        Arrays.fill(dp,-1);
        
        int ans = dfs(0,nums);
        return ans;
    }
}

File: 198-house-robber/house-robber.kt
class Solution {
    val vis = IntArray(101){-1}

    fun dfs(ind: Int, nums: IntArray): Int {
        if(ind >= nums.size){
            return 0;
        }
        if(vis[ind] != -1){
            return vis[ind];
        }
        //not rob
        val left = dfs(ind+1,nums);
        //rob
        val right =  nums[ind] + dfs(ind + 2,nums);

        val max = Math.max(left,right); 
        vis[ind] = max;
        return vis[ind];
    }
    fun rob(nums: IntArray): Int {
        return dfs(0,nums);
    }
}